lib/async_hooks.js

```diff
@@ -204,6 +205,105 @@ class AsyncResource {
   }
 }
 
+const locals = [];
+const asyncContextHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    for (let i = 0; i < locals.length; ++i) {
+      locals[i]._propagate(resource, executionAsyncResource());
+    }
+  }
+});
+
+class AsyncLocalStorage {
+  constructor() {
+    this.kResourceStore = Symbol('kResourceStore');
+  }
+
+  /**
+   * Propagate the context from a parent resource to a child one.
+   */
+  _propagate(resource, triggerResource) {
+    const store = triggerResource[this.kResourceStore];
+    if (store) {
+      resource[this.kResourceStore] = store;
+    }
+  }
+
+  _enter() {
+    asyncContextHook.enable();
+    if (!locals.includes(this)) {
```

Comment:
So, we need to have a structure keeping all the instances somewhere so the `init` hook can iterate over it. It used to be a `Set` which has better perf with `has` but we still need to loop around all instances when propagating.
Wdyt if I add a `Set` there and keep the array? We would loop over the array when propagating and use the `Set` when checking if the instance is already there.

Reply:
Would it be possible to refactor the logic so that we can remove `locals` entirely? I'm possibly lost on how that piece of logic work. Maybe some comments would help.

Reply:
Can you explain why you need to have a *global* data structure to iterate over? Cannot a resource-specific list be added to triggerResource for example?

Reply:
> Can you explain why you need to have a global data structure to iterate over? Cannot a resource-specific list be added to triggerResource for example?

you mean something like `triggerResource[kStorages] = [asyncLocalStorage1, asyncLocalStorage2]` ?
Indeed, we would not need `locals` anymore. It would make the addition of a `disable` flow a bit more complex as when going through `_propagrate` there would require a check on wether the instance is active or not. Also, this means that it will not be possible to disable the `AsyncHook` as there is no way to globally know how many instances are running. We could still have a global counter to that purpose however.

Reply:
Yes, exactly.

Reply:
Yes, exactly.

Reply:
Done in https://github.com/nodejs/node/pull/26540/commits/c996e7ee247e4636b7fba9b73f2f1105a330b9bd (along with introducing `enable/disable`. Let me know what you think.